pkgs/development/compilers/gcc/common/configure-flags.nix

```
@@ -111,9 +111,30 @@ let
       "--with-mpc=${libmpc}"
     ]
     ++ lib.optional (libelf != null) "--with-libelf=${libelf}"
-    ++ lib.optional (!(crossMingw && crossStageStatic))
+    ++ lib.optionals (!(crossMingw && crossStageStatic)) [
       "--with-native-system-header-dir=${lib.getDev stdenv.cc.libc}/include"
 
+      # gcc builds for cross-compilers (build != host) or cross-built
+      # gcc (host != target) always apply the offset prefix to disentangle
+      # target headers from build or host headers:
+      #     ${with_build_sysroot}${native_system_header_dir}
+      #  or ${test_exec_prefix}/${target_noncanonical}/sys-include
+      #  or ${with_sysroot}${native_system_header_dir}
+      # While native build (build == host == target) uses passed headers
+      # path as is:
+      #    ${native_system_header_dir}
+      #
+      # Nixpkgs uses flat directory structure for both native and cross
+      # cases. As a result libc headers don't get found for cross case
+      # and many modern features get disabled (libssp is used instead of
+      # target-specific implementations and similar). More details at:
+      #   https://github.com/NixOS/nixpkgs/pull/181802#issuecomment-1186822355
+      #
+      # We pick "/" path to effectively avoid sysroot offset and make it work
+      # as a native case.
+      "--with-build-sysroot=/"
```

```suggestion
    ] ++ lib.optionals (buildPlatform != hostPlatform) [
          "--with-build-sysroot=/"
```

Your original PR 4e7037168475458af2a4e30012f579adcda60ca9 had `lib.optionals (buildPlatform != hostPlatform ...`, but that conditional seems to be missing from dd1215865e1fccdedf099dd20cb0610a5c9cc8be.  Was that intentionanl?

---

I don't think my formal approval matters here, but this is my only remaining concern with this PR -- other than this (which I think was a typo?) everything LGTM.

Sorry for dragging out the review for so long.  Thank you again for untangling this!

---

Yeah, I intentionally removed extra check to always apply sysroot change wheh native-system-headers-dir is used. In `build == host != target)` case it was breaking in an unexpected way in `static-cross` case. But `--with-build-sysroot=/` it Just Works and should be a no-op.

---

> But `--with-build-sysroot=/` it Just Works and should be a no-op.

Let's hope so, because the new version of this PR is going to rebuild literally everything on Hydra. :astonished:

The old code and your original 4e70371 would send `build==host==target` through the `else` branch of [this if-then-else](https://gcc.gnu.org/git/?p=gcc.git;a=blob;f=gcc/configure.ac;h=446747311a6aec3c810ad6aa4190f7bd383b94f7;hb=HEAD#l2438).  Your latest dd12158 will send `build==host==target` through the `then` branch.  I've managed to convince myself that, for the `configureFlags` we use, those branches will both do the same thing as long as `sysroot_headers_suffix` is `""`.  I'm not really sure what "sysroot headers suffix" is for, I don't think I've ever encountered a system that used it.

That `if-then-else` block has been stable since gcc commit 0c90d48b346d, which was shipped in gcc 8.1.0.  So just to be totally completely sure:

@ofborg build gcc48 gcc49 gcc6 gcc7